homeassistant/components/heos/config_flow.py

```diff
@@ -267,6 +233,50 @@ async def async_step_reauth_confirm(
             ),
         )
 
+    async def _async_handle_discovered(self, hostname: str) -> ConfigFlowResult:
+        entry: HeosConfigEntry | None = await self.async_set_unique_id(DOMAIN)
+        # Abort early when discovery is ignored or host is part of the current system
+        if entry and (
+            entry.source == SOURCE_IGNORE or hostname in _get_current_hosts(entry)
+        ):
+            return self.async_abort(reason="single_instance_allowed")
+
+        # Connect to discovered host and get system information
+        heos = Heos(HeosOptions(hostname, events=False, heart_beat=False))
+        try:
+            await heos.connect()
+            system_info = await heos.get_system_info()
+        except HeosError as error:
+            _LOGGER.debug(
+                "Failed to retrieve system information from discovered HEOS device %s",
+                hostname,
+                exc_info=error,
+            )
+            return self.async_abort(reason="cannot_connect")
+        finally:
+            await heos.disconnect()
+
+        # Select the preferred host, if available
+        if system_info.preferred_hosts and system_info.preferred_hosts[0].ip_address:
```

Comment:
This method is cut/pasted / extracted from the original code (removed lines above). The only change is this line. `hostname` now is a parameter and declared as `str`, while `system_info.preferred_hosts[0].ip_address` can be `None`. That's why I added a check that it's actually set. This might be an `assert` as well in the `if`? Not sure if an entry in the list can exist while having no (/`None`) `ip_address`.